src/org/rascalmpl/repl/TerminalProgressBarMonitor.java

```diff
@@ -109,6 +117,96 @@ public void write(String str, int off, int len) throws IOException {
         }  
     }
 
+    private class UnfinishedLine {
+        final long threadId;
+        int curCapacity = 512;
+        byte[] buffer = new byte[curCapacity];
+        int curEnd = 0;
+        int lastNewLine = 0;
+
+        public UnfinishedLine() {
+            this.threadId = Thread.currentThread().getId();
+        }
+
+        /**
+         * Adding input combines previously unfinished sentences with possible
+         * new sentences. A number of cases come together here that otherwise
+         * should be diligently separated. 
+         * 
+         *   - An unfinished line can already exist or not
+         *   - The new input can contain newlines or not
+         *   - The new input can end with a newline character or not
+         * 
+         * By concatenating the previous unfinished line with the new input
+         * all we have to do now is figure out where the last newline character is.
+         * 
+         */
+        private void add(byte[] newInput, int offset, int len) {
+            // first ensure capacity of the array
+            if (curEnd + len >= curCapacity) {
+                var oldCapacity = curCapacity;
+                curCapacity *= 2; // this should not happen to often. we're talking a few lines of text.
+                byte[] tmp = new byte[curCapacity];
+                System.arraycopy(buffer, 0, tmp, 0, oldCapacity);
+                buffer = tmp;
+            }
+
+            System.arraycopy(newInput, offset, buffer, curEnd, len);
+            curEnd += len;
+            lastNewLine = startOfLastLine();
+        }
+
+        public void write(byte[] n, OutputStream out) throws IOException {
+            add(n, 0, n.length);
+            flushToLastLine(out);
+        }
+
+        public void write(byte[] n, int offset, int len, OutputStream out) throws IOException {
+            add(n, offset, len);
+            flushToLastLine(out);
+        }
+
+        private void flushToLastLine(OutputStream out) throws IOException {
+            if (lastNewLine != -1) {
+                // write everything out (except the last unfinished line), but including the last newline
+                out.write(buffer, 0, lastNewLine + 1);
+                
+                if (lastNewLine + 1 == curEnd) {
+                    // nothing left
+                    curEnd = 0;
+                    lastNewLine = -1;
+                }
+                else {
+                    // copy the last line that does not end with a newline
+                    curEnd -= lastNewLine;
+                    System.arraycopy(buffer, lastNewLine + 1, buffer, 0, curEnd);
+                    lastNewLine = -1; // no newline anymore
+                }
+            }
+
+            // otherwise we wait until the next input comes to be able to complete a line.
+        }
+
+        public void writeLeftOvers(OutputStream out) throws IOException {
```

Comment:
nit: rename to `flush`?

Reply:
It's not the same as a flush because it also adds newlines. So I have renamed it.